module/CMakeLists.txt

```diff
@@ -245,73 +268,59 @@ if(DEAL_II_WITH_CXX20_MODULE)
       )
 
+    #
     # Say what the source files for the library are. These are all of
     # the interface and implementation partition units we have created
     # above (including the ones that wrap external libraries), plus
-    # the primary interface unit. Some of these will have to be
-    # installed -- specifically, the interface and primary interface
-    # units -- and so need to be PUBLIC.  On the other hand,
-    # downstream projects do not need access to our implementation,
-    # and consequently the implementation units can be marked as
-    # PRIVATE. Finally, the macros.h header is also a PUBLIC source
-    # that needs to be installed.
+    # the primary interface unit.
     #
-    # TODO: It should not be necessary to list BASE_DIRS here. The module
-    # files are create in the build directory, and are all listed as absolute
-    # paths, not relative to a base dir. But cmake still complains when one
-    # omits the BASE_DIRS part, even though there is no reason for it to
-    # check whether these files can be found in the implicit BASE_DIRS (namely,
-    # CMAKE_CURRENT_SOURCE_DIR) because absolute paths to these files are
-    # provided.
+
     target_sources(${DEAL_II_TARGET_NAME}_module_${build_lowercase}
       PUBLIC
-      FILE_SET ${DEAL_II_TARGET_NAME}_module_${build_lowercase}_sources
+      FILE_SET module_interface_units
       TYPE CXX_MODULES
       FILES
         ${_interface_module_partition_units}
         ${_primary_module_interface_unit}
-      BASE_DIRS
-        ${CMAKE_CURRENT_SOURCE_DIR}
-        ${CMAKE_CURRENT_BINARY_DIR}
+      BASE_DIRS "${CMAKE_CURRENT_BINARY_DIR}/interface_module_partitions"
+      )
+    # Workaround an issue with CMake 4.0.3 (and possible earlier) where the
+    # export() logic loses one of the BASE_DIRS. Thus, record the following
+    # files separately
+    target_sources(${DEAL_II_TARGET_NAME}_module_${build_lowercase}
+      PUBLIC
+      FILE_SET external_module_interface_units
+      TYPE CXX_MODULES
+      FILES ${_external_interface_module_partition_units}
+      BASE_DIRS "${CMAKE_CURRENT_SOURCE_DIR}/interface_module_partitions"
       )
 
     target_sources(${DEAL_II_TARGET_NAME}_module_${build_lowercase}
       PRIVATE
-      FILE_SET ${DEAL_II_TARGET_NAME}_module_${build_lowercase}_sources_private
+      FILE_SET module_implementation_units
       TYPE CXX_MODULES
       FILES
         ${_implementation_module_partition_units}
-      BASE_DIRS
-        ${CMAKE_CURRENT_SOURCE_DIR}
-        ${CMAKE_CURRENT_BINARY_DIR}
+      BASE_DIRS ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR}
       )
 
     target_sources(${DEAL_II_TARGET_NAME}_module_${build_lowercase}
       PUBLIC
-      FILE_SET ${DEAL_II_TARGET_NAME}_module_${build_lowercase}_headers
+      FILE_SET headers
       TYPE HEADERS
-      FILES
-        ${_dealii_macros_header}
-      BASE_DIRS
-        ${CMAKE_CURRENT_BINARY_DIR}/../include/deal.II
+      FILES ${_dealii_macros_header}
```

Comment:
I take it that the semantics of `FILE_SET` here is that it *adds* the files to the existing file set handle `headers`? That would also explain why you can re-use the same name for both debug and release libraries above. Indeed, I find this:
```
FILE_SET <set>

The name of the file set to create or add to. It must contain only letters, numbers and underscores. Names starting with a capital letter are reserved for built-in file sets predefined by CMake. The only predefined set names are those matching the acceptable types. All other set names must not start with a capital letter or underscore.
```
Don't you love special rules such as upper/lower-case variable names...